src/inmanta/server/services/compilerservice.py

```diff
@@ -218,61 +218,73 @@ async def run(self, force_update: Optional[bool] = False) -> Tuple[bool, Optiona
-                await self._run_compile_stage("Updating modules", inmanta_path + ["modules", "update"], project_dir)
+                return await self._run_compile_stage(
+                    "Updating modules", [*inmanta_path, "-vvv", "-X", "modules", "update"], project_dir
+                )
 
-            async def install_modules() -> None:
+            async def install_modules() -> data.Report:
                 LOGGER.info("Installing modules")
-                await self._run_compile_stage(
+                return await self._run_compile_stage(
                     "Installing modules", [*inmanta_path, "-vvv", "-X", "project", "install"], project_dir
                 )
 
-            repo_url: str = env.repo_url
-            repo_branch: str = env.repo_branch
-            if os.path.exists(os.path.join(project_dir, "project.yml")):
-                await self._end_stage(0)
-
-                should_update: bool = force_update or self.request.force_update
-
-                # switch branches
-                if repo_branch:
-                    branch = await self.get_branch()
-                    if branch is not None and repo_branch != branch:
-                        if should_update:
-                            await self._run_compile_stage("Fetching new branch heads", ["git", "fetch"], project_dir)
-                        result = await self._run_compile_stage(
-                            f"Switching branch from {branch} to {repo_branch}", ["git", "checkout", repo_branch], project_dir
-                        )
-                        if not should_update:
-                            # if we update, update procedure will install modules
-                            await install_modules()
-
-                # update project
-                if should_update:
-                    await self._run_compile_stage("Pulling updates", ["git", "pull"], project_dir)
-                    await update_modules()
-            else:
-                if not repo_url:
-                    await self._warning(f"Failed to compile: no project found in {project_dir} and no repository set.")
-                    await self._end_stage(1)
-                    return False, None
-
-                if len(os.listdir(project_dir)) > 0:
-                    await self._warning(f"Failed to compile: no project found in {project_dir} but directory is not empty.")
-                    await self._end_stage(1)
-                    return False, None
-
-                await self._end_stage(0)
-
-                # clone repo and install project
-                cmd = ["git", "clone", repo_url, "."]
-                if repo_branch:
-                    cmd.extend(["-b", repo_branch])
-                result = await self._run_compile_stage("Cloning repository", cmd, project_dir)
-                if result.returncode is None or result.returncode > 0:
+            async def setup() -> AsyncIterator[Awaitable[data.Report]]:
```

Comment:
Why have this very exotic structure?

Reply:
Each of the `data.Report` instances needs their returncode checked. I initially added these checks after each call but it made the whole logic harder to read in my opinion. I could revert it to that if you prefer.